core/kernel/tcg.c

```diff
@@ -0,0 +1,393 @@
+	*pos += spec_event_size;
+
+	if (get_unaligned_le32(&event_header->event_size) != spec_event_size) {
+		EMSG("specID event: header event size mismatch\n");
+		/* Right way to handle this can be to call SetActive PCR's */
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * tcg_parse_event() -  Parse the event in the eventlog
+ *
+ * @buffer:		Pointer to the start of the eventlog
+ * @log_size:		Size of the eventlog
+ * @offset:		[in] Offset of the event in the eventlog buffer
+ *			[out] Return offset of the next event in the buffer
+ * @digest_list:	list of digests in the event
+ * @pcr			Index of the PCR in the event
+ *
+ * Return:		status code
+ */
+static TEE_Result tcg_parse_event(void *buffer, uint32_t log_size,
+				  uint32_t *offset,
+				  struct tpml_digest_values *digest_list,
+				  uint32_t *pcr)
+{
+	struct tcg_pcr_event2 *event = NULL;
+	uint32_t count = 0, size = 0, event_size = 0;
+	uint32_t i = 0;
+	size_t pos = 0;
+
+	event_size = tcg_event_final_size(digest_list);
+	if (*offset >= log_size || *offset + event_size > log_size) {
+		EMSG("Event exceeds log size\n");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	event = (struct tcg_pcr_event2 *)((uintptr_t)buffer + *offset);
+	*pcr = get_unaligned_le32(&event->pcr_index);
+
+	/* get the count */
+	count = get_unaligned_le32(&event->digests.count);
+	if (count != digest_list->count)
+		return TEE_ERROR_BAD_FORMAT;
+
+	pos = offsetof(struct tcg_pcr_event2, digests);
+	pos += offsetof(struct tpml_digest_values, digests);
+
+	for (i = 0; i < digest_list->count; i++) {
+		uint16_t alg;
+		uint16_t hash_alg = digest_list->digests[i].hash_alg;
+		uint8_t *digest = (uint8_t *)&digest_list->digests[i].digest;
+
+		alg = get_unaligned_le16((void *)((uintptr_t)event + pos));
+
+		if (alg != hash_alg)
+			return TEE_ERROR_BAD_FORMAT;
+
+		pos += offsetof(struct tpmt_ha, digest);
+		memcpy(digest, (void *)((uintptr_t)event + pos),
+		       tpm2_get_alg_len(hash_alg));
+		pos += tpm2_get_alg_len(hash_alg);
+	}
+
+	size = get_unaligned_le32((void *)((uintptr_t)event + pos));
+	event_size += size;
+	pos += sizeof(uint32_t); /* tcg_pcr_event2 event_size*/
+	pos += size;
+
+	/* make sure the calculated buffer is what we checked against */
+	if (pos != event_size)
+		return TEE_ERROR_BAD_FORMAT;
+
+	if (pos > log_size)
+		return TEE_ERROR_BAD_FORMAT;
+
+	*offset += pos;
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * tcg_process_fw_eventlog() - Parse the eventlog and extend the PCR's
+ *
+ * Return:	status code
+ */
+TEE_Result tcg_process_fw_eventlog(void)
+{
+	void *buffer = NULL;
+	void *tmp = NULL;
+	uint32_t i = 0, pcr = 0, pos = 0;
+	size_t digest_list_sz = 0;
+	size_t sz = TPM2_EVENT_LOG_SIZE;
+	TEE_Result ret = TEE_SUCCESS;
+	struct tpml_digest_values *digest_list = NULL;
+
+	if (!pcr_provider) {
+		EMSG("No provider available for PCR's");
+		return TEE_ERROR_GENERIC;
+	}
+
+	buffer = malloc(TPM2_EVENT_LOG_SIZE);
+	if (!buffer) {
+		EMSG("Error allocating mem\n");
```

Comment:
`\n` not needed.

Reply:
Sorry, I didn't notice it. I will remove it from here as well as other occurrence in the file for \n